Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +22,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHours_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(currentNew_ < 0)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+int BatteryStateOfChargeService::getRemainingHours() const
+{
+    return remainingHours_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return timeTillChargeOrDepletion_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = timeNew_;
+    double currentOld = currentNew_;
+
+    currentNew_ = batteryData.current;
+    timeNew_ = batteryData.time;
+
+    if(timeOld.isNull()) // first itteration
+    {
+        ampHours_ = BATTERY_AMP_HOUR_CAPACITY * initialStateOfChargePercent_ / PERCENT_TO_DECIMAL;
+    }
+    else
+    {
+        double avgCurrent = (currentOld + currentNew_) / 2;
+        ampHours_ += (avgCurrent * timeOld.secsTo(timeNew_) / SECONDS_TO_HOURS);
+    }
+
+    //calculate remaining time for charge or depletion
+    double totalHours;
+    int hours;
+    int minutes;
+    int seconds;
+
+    if(isCharging())
+    {
+        totalHours = qAbs(totalAmpHoursUsed() / currentNew_);
+    }
+    else
+    {
+        totalHours = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / currentNew_;
+    }
+
+    hours = (int)totalHours;
+    minutes = (int)((totalHours - hours) * HMS_CONVERSION_FACTOR);
+    seconds = (int)((totalHours * HMS_CONVERSION_FACTOR - hours * HMS_CONVERSION_FACTOR - minutes) * HMS_CONVERSION_FACTOR);
+
+    remainingHours_ = ((int)(totalHours / HOURS_IN_A_DAY)) * HOURS_IN_A_DAY;
```

Comment:
This comment will be miscellaneous from the above, but it looks like you could have used `remainingHours_ = hours % HOURS_IN_A_DAY`?

Reply:
remainingHours_ can be larger than 24